node.py

```diff
@@ -52,20 +55,28 @@ def init_coordinator_connection(self, coordinator_host, coordinator_port):
             raise e
 
     def init_etcd_connection(self, etcd_host, etcd_port):
-        """Initialize connection to etcd and write own metadata"""
+        """Initialize connection to etcd"""
         self.etcd_host = etcd_host
         self.etcd_port = etcd_port
 
         # self.etcd_client is the inbuilt database connection for node metadata
         self.etcd_client = etcd3.client(host=etcd_host, port=etcd_port)
-
-        # write own metadata to etcd
+            
+    def claim_port(self):
+        """Claims a port by creating a lease with etcd, and adds the node metadata"""
         try:
+            self.port = network.find_available_port(self.etcd_client, self.host)
+
+            self.lease = self.etcd_client.lease(ttl=etcd_utils.ETCD_LEASE_EXPIRY_TIME)
+
+            self.etcd_client.put(f"ports/{self.host}/{self.port}", self.id, lease=self.lease)
```

Comment:
Edge case here. 
1. Say initially two process P1 and P2 calls `find_available_port` at the same time. 
2. Since no port is currently leased, its possible `find_available_port` both returns 1111 to P1 and P2. 
3. P2 then act slowly for wtv reason allowing P1 to grab 1111 first. 
4. P2 then try to lease 1111 as well but fails, and thus failed to acquire a port

Fixes for this
- claim_port should be made into a while loop to guarantee the process always claim some port (busy-wait but with random port choosing it shouldn't be bad
- some serialization  mechanism at `find_available_port` and lease a port asap before return back to `claim_port`